### PR TITLE
Print absolute path of TOML file, when parsing fails

### LINKED ISSUE #20368: Provide explicit path on parse error from toml Version Catalog file
When a user includes invalid syntax in a toml file, the error message currently produced does not have an explicit reference to the file in which the invalid syntax was found (though it has line and column number).

### Expected Behavior

The error message should include an absolute path to the file with a parse error.

### Current Behavior

If I define a project with its version catalog in `gradle/foo.versions.toml`, set up with
```
dependencyResolutionManagement {
    versionCatalogs {
        libs {
            from(files("gradle/foo.versions.toml"))
        }
    }
}
```
in `settings.gradle`, and make a syntax error in the Toml file (for example by attempting to comment out a line with `//` rather than `#`), the error message is:
```
* What went wrong:
org.gradle.api.InvalidUserDataException: Invalid TOML catalog definition:
  - Problem: In version catalog libs, parsing failed with 1 error.
    
    Reason: At line 2, column 1: Unexpected '/', expected a newline or end-of-input.
    
    Possible solution: Fix the TOML file according to the syntax described at https://toml.io.
    
    Please refer to https://docs.gradle.org/7.4/userguide/version_catalog_problems.html#toml_syntax_error for more details about this problem.
> Invalid TOML catalog definition:
    - Problem: In version catalog libs, parsing failed with 1 error.
      
      Reason: At line 2, column 1: Unexpected '/', expected a newline or end-of-input.
      
      Possible solution: Fix the TOML file according to the syntax described at https://toml.io.
      
      Please refer to https://docs.gradle.org/7.4/userguide/version_catalog_problems.html#toml_syntax_error for more details about this problem.
```

Nothing in the error message indicates that the problem was in `gradle/foo.versions.toml`.

### Context

Not having the absolute path in the error message makes it impossible for Android Studio to automatically open the file with the erroneous syntax, or to provide quick fixes or other actions in the IDE.

### PR DESCRIPTION
**Pull request by hegyibalint**
Fixes #20368

### CODE REVIEW

**Comment by hegyibalint** [comment]
@bot-gradle test RFM

**Comment by hegyibalint** [comment]
TL;DR of this, apart from the issue:

I've decided to refactor the model loading into a file-based approach, instead of using an input stream. 
Meanwhile, there is nothing wrong with using a stream, I can see that:
 - We are focused on loading files with the TOML parser, so it's improbable that we would need to change the code back
 - With the path-based approach, lower levels can easily know where the problem was originating from (the core of the issue being solved)

**Comment by hegyibalint** [comment]
@bot-gradle test RFM

**Comment by hegyibalint** [comment]
@bot-gradle test RFM

**Comment by hegyibalint** [comment]
@bot-gradle test RFM

**Comment by hegyibalint** [comment]
@bot-gradle test RFM

File: subprojects/plugins/src/test/groovy/org/gradle/api/plugins/catalog/internal/TomlWriterTest.groovy [hunk 1]
```diff
@@ -66,25 +71,46 @@ format.version = "1.1"
...
             Interners.newStrongInterner(),
             TestUtil.objectFactory(),
-            TestUtil.providerFactory()
-            ,
+            TestUtil.providerFactory(),
             Stub(Supplier))
-        ins.withCloseable {
-            TomlCatalogFileParser.parse(it, builder)
+
+        def inputStream = Files.newInputStream(path)
```
Comment (eskatos) [inline subprojects/plugins/src/test/groovy/org/gradle/api/plugins/catalog/internal/TomlWriterTest.groovy:111]: :question: is the input stream still necessary here?

File: subprojects/core/src/main/java/org/gradle/api/internal/catalog/parser/TomlCatalogFileParser.java [hunk 2]
```diff
@@ -79,31 +81,33 @@ public class TomlCatalogFileParser {
...
-        TomlTable versionsTable = result.getTable(VERSIONS_KEY);
-        TomlTable pluginsTable = result.getTable(PLUGINS_KEY);
-        Sets.SetView<String> unknownTle = Sets.difference(result.keySet(), TOP_LEVEL_ELEMENTS);
-        if (!unknownTle.isEmpty()) {
-            throwVersionCatalogProblem(builder, VersionCatalogProblemId.TOML_SYNTAX_ERROR, spec ->
-                spec.withShortDescription(() -> "Unknown top level elements " + unknownTle)
-                    .happensBecause(() -> "TOML file contains an unexpected top-level element")
-                    .addSolution(() -> "Make sure the top-level elements of your TOML file is one of " + oxfordListOf(TOP_LEVEL_ELEMENTS, "or"))
-                    .documented());
+        try (InputStream inputStream = Files.newInputStream(catalogFilePath)) {
```
Comment (eskatos) [inline subprojects/core/src/main/java/org/gradle/api/internal/catalog/parser/TomlCatalogFileParser.java:86]: :question: shouldn't this use a buffered stream? I don't know if `Toml` itself does any buffering.

File: subprojects/core/src/main/java/org/gradle/configuration/BuildTreePreparingProjectsPreparer.java [hunk 3]
```diff
@@ -75,6 +75,7 @@ private void generateDependenciesAccessorsAndAssignPluginVersions(ServiceRegistr
         dm.getDefaultLibrariesExtensionName().finalizeValue();
         String defaultLibrary = dm.getDefaultLibrariesExtensionName().get();
         File dependenciesFile = new File(settings.getSettingsDir(), "gradle/libs.versions.toml");
+        // TODO:configuration-cache causes test failures (see "configuration cache for help on empty project")
```
Comment (eskatos) [inline subprojects/core/src/main/java/org/gradle/configuration/BuildTreePreparingProjectsPreparer.java:78]: :question: I'm not sure I understand, could you clarify that comment?

**Review by eskatos — COMMENTED** [review]
It's looking good!

Mostly left some questions

File: subprojects/core/src/main/java/org/gradle/api/internal/catalog/parser/TomlCatalogFileParser.java (hunk 2 above)
Comment (eskatos) [reply subprojects/core/src/main/java/org/gradle/api/internal/catalog/parser/TomlCatalogFileParser.java:86]: Toml does buffer the stream!
Marking this question as resolved.

File: subprojects/core/src/main/java/org/gradle/configuration/BuildTreePreparingProjectsPreparer.java (hunk 3 above)
Comment (hegyibalint) [reply subprojects/core/src/main/java/org/gradle/configuration/BuildTreePreparingProjectsPreparer.java]: I'm not even sure that this line is right... we've been digging around the version catalogs with Rodrigo, and found some issues (e.g. here, where CC doesn't know that we try to open the default VC).

I think I will try to instrument this, and see if it behaves as expected.

File: subprojects/core/src/main/java/org/gradle/configuration/BuildTreePreparingProjectsPreparer.java [hunk 4]
```diff
@@ -75,6 +76,8 @@ private void generateDependenciesAccessorsAndAssignPluginVersions(ServiceRegistr
         dm.getDefaultLibrariesExtensionName().finalizeValue();
         String defaultLibrary = dm.getDefaultLibrariesExtensionName().get();
         File dependenciesFile = new File(settings.getSettingsDir(), "gradle/libs.versions.toml");
+        // Notify the CC
+        Instrumented.fileOpened(dependenciesFile, getClass().getName());
```
Comment (ljacomet) [inline subprojects/core/src/main/java/org/gradle/configuration/BuildTreePreparingProjectsPreparer.java:80]: Should that happen even when the file does not exist?

File: subprojects/workers/src/main/java/org/gradle/workers/internal/WorkerDaemonServer.java [hunk 5]
```diff
@@ -113,7 +113,7 @@ public String toString() {
 
     private static class WorkerDaemonServices extends WorkerSharedUserHomeScopeServices {
 
-        // TODO: configuration-cache - deprecate workers access to ProviderFactory?
+        // TODO:configuration-cache - deprecate workers access to ProviderFactory?
```
Comment (ljacomet) [inline subprojects/workers/src/main/java/org/gradle/workers/internal/WorkerDaemonServer.java:116]: I feel like this TODO is either not necessary or missing a real issue reference. Otherwise it has a high chance of becoming rot ...

**Review by ljacomet — CHANGES_REQUESTED** [review]
Nice improvement to the error message!

Some questions mostly around CC handling - see below

File: subprojects/core/src/main/java/org/gradle/configuration/BuildTreePreparingProjectsPreparer.java (hunk 4 above)
Comment (bamboo) [reply subprojects/core/src/main/java/org/gradle/configuration/BuildTreePreparingProjectsPreparer.java:80]: > Should that happen even when the file does not exist?

Should the configuration phase re-run when the file is later added? If so then yes.

File: subprojects/workers/src/main/java/org/gradle/workers/internal/WorkerDaemonServer.java (hunk 5 above)
Comment (bamboo) [reply subprojects/workers/src/main/java/org/gradle/workers/internal/WorkerDaemonServer.java:116]: It's ok to leave it as we screen the codebase for those every now and then.

File: subprojects/core/src/main/java/org/gradle/configuration/BuildTreePreparingProjectsPreparer.java (hunk 4 above)
Comment (eskatos) [reply subprojects/core/src/main/java/org/gradle/configuration/BuildTreePreparingProjectsPreparer.java:80]: Yes, it should be `fileObserver()` here and `fileOpened` in the builder.

**Comment by hegyibalint** [comment]
@bot-gradle test RFM please

File: subprojects/core/src/main/java/org/gradle/configuration/BuildTreePreparingProjectsPreparer.java (hunk 4 above)
Comment (hegyibalint) [reply subprojects/core/src/main/java/org/gradle/configuration/BuildTreePreparingProjectsPreparer.java]: Changed `fileOpened` to `fileObserved`.

**Review by ljacomet — APPROVED** [review]
